GEECS-PythonAPI/geecs_python_api/analysis/scans/scan_data.py

```
@@ -25,15 +25,22 @@
 if not logging.getLogger().hasHandlers():
     logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s")
 
-try:
-    CONFIG = GeecsPathsConfig()
-except ValueError:
-    logging.error("'CONFIG' not set for ScanData")
-    CONFIG = None
+
+def reload_paths_config():
+    """ Used by GEECS Scanner to fix scan_data_manager in case experiment name has changed """
+    try:
+        ScanData.paths_config = GeecsPathsConfig()
+    except ValueError:
+        logger.error("'CONFIG' not set for ScanData")
+        ScanData.paths_config = None
+
 
 class ScanData:
     """ Represents a GEECS experiment scan """
 
+    paths_config = None  # TODO the "reload" function is called statically at the bottom of this file, seems kinda bad
```

The reload_paths_config() function could be a classmethod (not to be confused with staticmethod).
Also consider changing the behavior of GeecsPathsConfig(). It's very strange for an `__init__` method to throw a ValueError.

---

Oooh interesting!  I've never seen a classmethod before, but yeah seems like the perfect use case for it.

And yeah I'll change the ValueError to something better, I remember back when I was browsing the python errors this seemed like the best one at the time.  I think what I'll do is make a new exception class and use that.

---

Oh I meant I don't think an `__init__` function should throw an error at all, except maybe validation of constructor arguments. Especially in this case, where GeecsPathsConfig is dataclass-like. 

There are a few options: 
* Add a load() method, to be called by the caller, who can catch errors.
* Replace raising ValueError by setting things to None, and then check for that in the caller. 
* Add a staticmethod load() that returns a GeecsPathsConfig. 

And finally, after thinking about it a bit more and reading a bit, it's probably ok just to leave it as is - it's the simplest and cleanest, and it isn't bad form like I thought before.

---

Ok, I think I see what you're saying.  I suppose if `paths_config` were always a GEECSPathsConfig class instance then it moves all of the error handling when not properly initialized to the class functions of GEECSPathsConfig rather than in ScanData or anywhere else using it.  And this would probably be cleaner if more modules were using this class.

I'll keep this in mind, the refactoring would not be too expansive I think.  But I'll save this for another day.